### PR TITLE
Fix: V1.12 RO-4033: Adjust logic used for processing and assigning sequence alignment details

### PR DESCRIPTION
**Pull request by piehld**
Jira ticket: https://rcsbpdb.atlassian.net/browse/RO-4033

Address issue observed for 2B7X_1, described here: https://rcsbpdb.atlassian.net/browse/RO-4033?focusedCommentId=221428

> The problem does relate to ExDB logic, but I think there is a reason the logic is written the way it is. Let me explain what happens:
> 
> First, the code looks up the SIFTS UniProt-mapping/alignment data for all chains in the given entity, in this case 2B7X_1, which has 4 identical chains: A, B, C, D. The data is as follows:
> ```
> Chains A, C, and D have the same alignment details:
> [{'UP': 'P00720', 'BG': 1, 'LEN': 30, 'UBG': 1, 'UND': 30}, {'UP': 'P00720', 'BG': 37, 'LEN': 134, 'UBG': 31, 'UND': 164}]
> 
> Chain B has an alternate alignment:
> [{'UP': 'P00720', 'BG': 1, 'LEN': 24, 'UBG': 1, 'UND': 24}, {'UP': 'P00720', 'BG': 31, 'LEN': 140, 'UBG': 25, 'UND': 164}]
> ```
> Then, the code pools these alignment segments together…
> ```
> [
>  {'UP': 'P00720', 'BG': 1, 'LEN': 30, 'UBG': 1, 'UND': 30}      # Chain A
>  {'UP': 'P00720', 'BG': 37, 'LEN': 134, 'UBG': 31, 'UND': 164}  # Chain A
>  {'UP': 'P00720', 'BG': 1, 'LEN': 24, 'UBG': 1, 'UND': 24}      # Chain B
>  {'UP': 'P00720', 'BG': 31, 'LEN': 140, 'UBG': 25, 'UND': 164}  # Chain B
>  {'UP': 'P00720', 'BG': 1, 'LEN': 30, 'UBG': 1, 'UND': 30}      # Chain C
>  {'UP': 'P00720', 'BG': 37, 'LEN': 134, 'UBG': 31, 'UND': 164}  # Chain C
>  {'UP': 'P00720', 'BG': 1, 'LEN': 30, 'UBG': 1, 'UND': 30}      # Chain D
>  {'UP': 'P00720', 'BG': 37, 'LEN': 134, 'UBG': 31, 'UND': 164}  # Chain D
> ]
> ```
> and then splits/groups them into non-overlapping segments:
> ```
> 1: [
>     {'UP': 'P00720', 'BG': 1, 'LEN': 30, 'UBG': 1, 'UND': 30}      # Chain A  *
>     {'UP': 'P00720', 'BG': 1, 'LEN': 24, 'UBG': 1, 'UND': 24}      # Chain B
>     {'UP': 'P00720', 'BG': 1, 'LEN': 30, 'UBG': 1, 'UND': 30}      # Chain C
>     {'UP': 'P00720', 'BG': 1, 'LEN': 30, 'UBG': 1, 'UND': 30}      # Chain D
> 2: [
>     {'UP': 'P00720', 'BG': 37, 'LEN': 134, 'UBG': 31, 'UND': 164}  # Chain A
>     {'UP': 'P00720', 'BG': 31, 'LEN': 140, 'UBG': 25, 'UND': 164}  # Chain B  *
>     {'UP': 'P00720', 'BG': 37, 'LEN': 134, 'UBG': 31, 'UND': 164}  # Chain C
>     {'UP': 'P00720', 'BG': 37, 'LEN': 134, 'UBG': 31, 'UND': 164}  # Chain D
> ]
> ```
> Last, the code picks the segment from each group with the longest length (LEN), which I’ve indicated with the asterisk above. As a result, we end up with the problematic combination of aligned segments.
> 
> This logic seems to try to handle cases where different chains of the same entity demonstrate different extents of alignment (e.g., if chain A has coverage near the N-terminus, while chain B has coverage near the C-terminus), or the possibility of chimeric entities. Perhaps you see another situation that this logic was intended to address. 
> 
> In order to adjust this logic to handle the case of 2B7X (and others that are similarly-affected), the most obvious solution I see would be to keep the alignment segments grouped by chain, and then pick the set of aligned segments that has the most total coverage.

### CODE REVIEW

File: rcsb/utils/dictionary/DictMethodEntityHelper.py [hunk 1]
```diff
@@ -97,13 +98,26 @@ def __processSiftsAlignments(self, dataContainer):
...
+            # Only keep the chain with the longest alignment
+            if asaolenLength > asymMaxAlignLength:
+                siftsAlignD[(entryId, entityId)] = asymSeqAlignObjL
+                logger.debug("siftsAlignD: %r", siftsAlignD)
+                asymMaxAlignLengthD[(entryId, entityId)] = asaolenLength
+            #
             # siftsAlignD.setdefault((entryId, entityId), []).extend([SeqAlign("SIFTS", **sa) for sa in self.__ssP.getIdentifiers(entryId, authAsymId, idType="UNPAL")])
-            siftsAlignD.setdefault((entryId, entityId), []).extend(self.__ssP.getSeqAlignObjList(entryId, authAsymId))
+
         for (entryId, entityId), seqAlignObjL in siftsAlignD.items():
```
Comment (brindakv) [inline rcsb/utils/dictionary/DictMethodEntityHelper.py:121]: With the proposed changes, the code segment in lines 129-136 now seems to be unnecessary based on the new additions in lines 107-117. 

If I understand this correctly (and please correct me if I am wrong), identifying the longest overlapping entity region only applies if there is more than one alignment between an entity and a reference sequence, which happens with homomeric structures. But that logic is now addressed earlier when only the longest alignment for an entity is selected. So the code segment (lines 129-136) seems redundant.

**Review by brindakv — COMMENTED** [review]
Thanks @piehld. Please see my comment below.

File: rcsb/utils/dictionary/DictMethodEntityHelper.py (hunk 1 above)
Comment (piehld) [reply rcsb/utils/dictionary/DictMethodEntityHelper.py:121]: Thank you Brinda, you are correct that was redundant code. It was basically rewrangling the data into the a new dictionary and in the same structure. I've consolidated the code and just finished testing it over a full reload, which appeared to work for the `CoreDataLoading` task (as the alignment was then fixed).

However, upon running the next task of the workflow, `ExdbDataLoading`, I discovered that task ends up overwriting the alignments by the [ReferenceSequenceAssignmentAdapter.py](https://github.com/rcsb/py-rcsb_exdb/blob/556f3491c8df14c45059218fcf8e1a86df58c3f4/rcsb/exdb/seq/ReferenceSequenceAssignmentAdapter.py#L311C20-L311C20) process. So now, the alignments are reverted back to the incorrect values. Accordingly, I'll need to update the code in the py-rcsb_exdb package to resolve that.

Comment (piehld) [reply rcsb/utils/dictionary/DictMethodEntityHelper.py:121]: OK, I've implemented and tested a fix here: https://github.com/rcsb/py-rcsb_utils_seq/pull/15

Comment (brindakv) [reply rcsb/utils/dictionary/DictMethodEntityHelper.py:121]: Thanks @piehld. Looks good to me.

**Review by brindakv — APPROVED** [review]

**Pull request merged**